File: packages/shared/src/connectors/fearGreed.ts
import { fetchJson } from "./base.js";

// alternative.me's Fear & Greed Index -- free, keyless, no rate-limit
// concerns observed. Confirmed live before use (curl, no API key needed).
// Market-wide, not per-token: blends volatility, momentum/volume, social
// volume, surveys, dominance, and trends into one daily 0-100 number, per
// their own published methodology. This is NOT itself a prediction -- it's
// a real-time snapshot of aggregate market mood, same category as
// CoinGecko's own sentiment_votes_up/down_percentage already used elsewhere
// in this app.
const FNG_URL = "https://api.alternative.me/fng/?limit=1";

export interface FearGreedIndex {
  value: number;
  classification: string;
  updatedAt: string;
}

interface FearGreedResponse {
  data: { value: string; value_classification: string; timestamp: string }[];
}

export async function fetchFearGreedIndex(): Promise<FearGreedIndex | null> {
  const data = await fetchJson<FearGreedResponse>(FNG_URL);
  const entry = data?.data?.[0];
  if (!entry) return null;
  return {
    value: Number(entry.value),
    classification: entry.value_classification,
    updatedAt: new Date(Number(entry.timestamp) * 1000).toISOString(),
  };
}

File: mobile/lib/connectors/fear_greed.dart
// Mirrors packages/shared/src/connectors/fearGreed.ts -- alternative.me's
// Fear & Greed Index, free and keyless, confirmed live before use. Market-
// wide, not per-token; see the web connector's own comment for the "this
// isn't a prediction" reasoning (same category as CoinGecko's own
// sentiment_votes_up/down_percentage, already used elsewhere in this app).
import 'http_client.dart';

const _fngUrl = 'https://api.alternative.me/fng/?limit=1';

class FearGreedIndex {
  final int value;
  final String classification;
  final String updatedAt;

  FearGreedIndex({required this.value, required this.classification, required this.updatedAt});
}

Future<FearGreedIndex?> fetchFearGreedIndex() async {
  final data = await fetchJson(_fngUrl);
  if (data == null) return null;
  final entries = data['data'] as List?;
  if (entries == null || entries.isEmpty) return null;
  final entry = entries.first as Map<String, dynamic>;
  final timestampSeconds = int.tryParse(entry['timestamp'] as String? ?? '');
  return FearGreedIndex(
    value: int.tryParse(entry['value'] as String? ?? '') ?? 0,
    classification: entry['value_classification'] as String? ?? 'Unknown',
    updatedAt: timestampSeconds != null
        ? DateTime.fromMillisecondsSinceEpoch(timestampSeconds * 1000).toIso8601String()
        : DateTime.now().toIso8601String(),
  );
}
